libxrdp/xrdp_sec.c

```diff
@@ -2302,160 +2303,256 @@ xrdp_sec_process_mcs_data_channels(struct xrdp_sec *self, struct stream *s)
-            client_info->height > 0x7FFE || client_info->height < 0xC8)
+
+    /* make sure virtual desktop size is ok - 2.2.1.3.6 Client Monitor Data (TS_UD_CS_MONITOR) */
+    if (description->session_width > CLIENT_MONITOR_DATA_MAXIMUM_VIRTUAL_DESKTOP_WIDTH || description->session_width < CLIENT_MONITOR_DATA_MINIMUM_VIRTUAL_DESKTOP_WIDTH ||
+            description->session_height > CLIENT_MONITOR_DATA_MAXIMUM_VIRTUAL_DESKTOP_HEIGHT || description->session_height < CLIENT_MONITOR_DATA_MINIMUM_VIRTUAL_DESKTOP_HEIGHT)
     {
         LOG(LOG_LEVEL_ERROR,
-            "Client supplied virtual desktop width or height is invalid. "
+            "\tprocess_monitor_stream: Client supplied virtual desktop width or height is invalid. "
             "Allowed width range: min %d, max %d. Width received: %d. "
             "Allowed height range: min %d, max %d. Height received: %d",
-            0xC8, 0x7FFE, client_info->width,
-            0xC8, 0x7FFE, client_info->height);
-        return 1; /* error */
+            CLIENT_MONITOR_DATA_MINIMUM_VIRTUAL_DESKTOP_WIDTH, CLIENT_MONITOR_DATA_MAXIMUM_VIRTUAL_DESKTOP_WIDTH, description->session_width,
+            CLIENT_MONITOR_DATA_MINIMUM_VIRTUAL_DESKTOP_HEIGHT, CLIENT_MONITOR_DATA_MAXIMUM_VIRTUAL_DESKTOP_HEIGHT, description->session_width);
+        return 1;
     }
 
     /* keep a copy of non negative monitor info values for xrdp_wm usage */
-    for (index = 0; index < monitorCount; index++)
+    for (int monitor_index = 0; monitor_index < num_monitor; ++monitor_index)
+    {
+        monitor_layout = description->minfo_wm + monitor_index;
+
+        g_memcpy(monitor_layout, description->minfo + monitor_index, sizeof(struct monitor_info));
+
+        monitor_layout->left = monitor_layout->left - all_monitors_encompassing_bounds.left;
+        monitor_layout->top = monitor_layout->top - all_monitors_encompassing_bounds.top;
+        monitor_layout->right = monitor_layout->right - all_monitors_encompassing_bounds.left;
+        monitor_layout->bottom = monitor_layout->bottom - all_monitors_encompassing_bounds.top;
+    }
+    return 0;
+}
+
+/*****************************************************************************/
+/* Process a [MS-RDPBCGR] TS_UD_CS_MONITOR message.
+   reads the client monitors data */
+static int
+xrdp_sec_process_mcs_data_monitors(struct xrdp_sec *self, struct stream *s)
+{
+    int flags;
+    struct xrdp_client_info *client_info = &(self->rdp_layer->client_info);
+
+    LOG_DEVEL(LOG_LEVEL_TRACE, "xrdp_sec_process_mcs_data_monitors:");
+
+    /* this is an option set in xrdp.ini */
+    if (client_info->multimon != 1) /* are multi-monitors allowed ? */
+    {
+        LOG(LOG_LEVEL_INFO, "\txrdp_sec_process_mcs_data_monitors: Multi-monitor is disabled by server config");
+        return 0;
+    }
+    if (!s_check_rem_and_log(s, 8, "\txrdp_sec_process_mcs_data_monitors: Parsing [MS-RDPBCGR] TS_UD_CS_MONITOR"))
```

Comment:
You're checking for 8 bytes on the stream here, of which 4 are used for the flags, and 4 are used for the monitor_count in `libxrdp_process_monitor_stream()`.  That works here, but in `xrdp_mm.c` your call to `libxrdp_process_monitor_stream()` is reading the stream without a check that the data is available (if I'm reading it right). Suggest you make this check for 4 bytes for the flags, and add a 4-byte check to `libxrdp_process_monitor_stream()` before reading the monitor count.

Hope that makes sense

Reply:
Done